src/main/java/com/microsoft/sqlserver/jdbc/IOBuffer.java

```diff
@@ -3574,27 +3596,81 @@ void writeTime(java.sql.Timestamp value, int scale) throws SQLServerException {
+        /*
+         * Out of all the supported temporal datatypes, DateTimeOffset is the only datatype that doesn't
+         * allow direct casting from java.sql.timestamp (which was created from a String).
+         * DateTimeOffset was never required to be constructed from a String, but with the
+         * introduction of extended bulk copy support for Azure DW, we now need to support this scenario.
+         * Parse the DTO as string if it's coming from a CSV.
+         */
+        if (value instanceof String) {
+            String stringValue = (String) value;
+            int lastColon = stringValue.lastIndexOf(':');
+
+            String offsetString = stringValue.substring(lastColon - 3);
+            minutesOffset = 60 * Integer.valueOf(offsetString.substring(1, 3))
+                    + Integer.valueOf(offsetString.substring(4, 6));
 
-        // If the target data type is DATETIMEOFFSET, then use UTC for the calendar that
-        // will hold the value, since writeRPCDateTimeOffset expects a UTC calendar.
-        // Otherwise, when converting from DATETIMEOFFSET to other temporal data types,
-        // use a local time zone determined by the minutes offset of the value, since
-        // the writers for those types expect local calendars.
-        timeZone = (SSType.DATETIMEOFFSET == destSSType) ? UTC.timeZone
-                                                         : new SimpleTimeZone(minutesOffset * 60 * 1000, "");
+            if (offsetString.startsWith("-"))
+                minutesOffset = -minutesOffset;
 
-        calendar = new GregorianCalendar(timeZone, Locale.US);
-        calendar.setLenient(true);
-        calendar.clear();
-        calendar.setTimeInMillis(utcMillis);
+            /*
+             * If the target data type is DATETIMEOFFSET, then use UTC for the calendar that
+             * will hold the value, since writeRPCDateTimeOffset expects a UTC calendar.
+             * Otherwise, when converting from DATETIMEOFFSET to other temporal data types,
+             * use a local time zone determined by the minutes offset of the value, since
+             * the writers for those types expect local calendars.
+             */
+            timeZone = (SSType.DATETIMEOFFSET == destSSType) ? UTC.timeZone
+                                                             : new SimpleTimeZone(minutesOffset * 60 * 1000, "");
+
+            calendar = new GregorianCalendar(timeZone, Locale.US);
+
+            String timestampString = stringValue.substring(0, lastColon - 4);
+            int year = Integer.valueOf(timestampString.substring(0, 4));
+            int month = Integer.valueOf(timestampString.substring(5, 7));
+            int day = Integer.valueOf(timestampString.substring(8, 10));
+            int hour = Integer.valueOf(timestampString.substring(11, 13));
+            int minute = Integer.valueOf(timestampString.substring(14, 16));
+            int second = Integer.valueOf(timestampString.substring(17, 19));
+
```

Comment:
why define all these variables they're only used once below

Reply:
I think it's easier to see this way.